business_requirement/models/business.py

```diff
@@ -0,0 +1,436 @@
+    confirmed_id = fields.Many2one(
+        'res.users', string='Confirmed by',
+        readonly=True
+    )
+    approval_date = fields.Datetime(
+        string='Approval Date',
+        readonly=True
+    )
+    approved_id = fields.Many2one(
+        'res.users',
+        string='Approved by',
+        readonly=True
+    )
+    company_id = fields.Many2one(
+        'res.company', string='Company',
+        required=True, readonly=True, states={'draft': [('readonly', False)]},
+        default=_get_default_company,
+    )
+
+    @api.one
+    @api.onchange('project_id')
+    def project_id_change(self):
+        if self.project_id and self.project_id.partner_id:
+            self.partner_id = self.project_id.partner_id.id
+
+    @api.model
+    def create(self, vals):
+        if vals.get('name', '/') == '/':
+            vals['name'] = self.env['ir.sequence'].get('business.requirement')
+        return super(BusinessRequirement, self).create(vals)
+
+    @api.multi
+    @api.depends('parent_id')
+    def _get_level(self):
+        def _compute_level(br):
+            return br.parent_id and br.parent_id.level + 1 or 1
+
+        for br in self:
+            level = _compute_level(br)
+            br.level = level
+
+    @api.one
+    @api.depends('business_requirement_ids')
+    def _sub_br_count(self):
+        self.sub_br_count = len(self.business_requirement_ids)
+
+    @api.model
+    def _get_states(self):
+        states = [
+            ('draft', 'Draft'),
+            ('confirmed', 'Confirmed'),
+            ('approved', 'Approved'),
+            ('customer_approval', 'Customer Approval'),
+            ('in_progress', 'In progress'),
+            ('done', 'Done'),
+            ('cancel', 'Cancel'),
+            ('drop', 'Drop'),
+        ]
+        return states
+
+    @api.one
+    def action_button_confirm(self):
+        self.write({'state': 'confirmed'})
+        self.confirmed_id = self.env.user
+        self.confirmation_date = fields.Datetime.now()
+
+    @api.one
+    def action_button_back_draft(self):
+        self.write({'state': 'draft'})
+        self.confirmed_id = self.approved_id = []
+        self.confirmation_date = self.approval_date = ''
+
+    @api.one
+    def action_button_approve(self):
+        self.write({'state': 'approved'})
+        self.approved_id = self.env.user
+        self.approval_date = fields.Datetime.now()
+
+    @api.one
+    def action_button_customer_approval(self):
+        self.write({'state': 'customer_approval'})
+
+    @api.one
+    def action_button_in_progress(self):
+        self.write({'state': 'in_progress'})
+
+    @api.one
+    def action_button_done(self):
+        self.write({'state': 'done'})
+
+    @api.one
+    def action_button_cancel(self):
+        self.write({'state': 'cancel'})
+
+    @api.one
+    def action_button_drop(self):
+        self.write({'state': 'drop'})
+
+    @api.cr_uid_ids_context
+    def message_post(self, cr, uid, thread_id, body='', subject=None,
```

Comment:
in v8 this is pending rewriting/refactoring (currently not good to rewrite the whole function in this module)